megamek/src/megamek/common/util/MegaMekFile.java

```diff
@@ -1,43 +1,179 @@
  * opened.  However, if it doesn't exist, then the file is opened from MM's
  * install directory instead.
  * 
  * @author arlith
  *
  */
 public class MegaMekFile {
-    
-    File file;
-    
+    private List<File> fileList = new ArrayList<>();
+    private List<File> subdirectoryList = new ArrayList<>();
+    private boolean isDirectory;
+    private static MMLogger logger = DefaultMmLogger.getInstance();
+
     public MegaMekFile(File parent, String child) {
         this(new File(parent, child).toString());
     }
-    
-    public MegaMekFile(String pathname) {
-        File userdataVersion = new File(Configuration.userdataDir(), pathname);
-        if (userdataVersion.exists()) {
-            file = userdataVersion;
+
+    public MegaMekFile(String pathName) {
+        File standardVersion = new File(pathName);
+        File userDataVersion = new File(Configuration.userdataDir(), pathName);
+        isDirectory = standardVersion.isDirectory();
+
+        if (isDirectory) {
+            // Create an array of files located in the standard file directory
+            List<File> standardFiles = splitFilesAndSubdirectories(standardVersion.listFiles(), true);
+
+            if (userDataVersion.exists()) {
+                // If we have a user data version of the directory, then we need to do some additional
+                // processing based on the files located within it
+                List<File> userDataSubdirectories = splitFilesAndSubdirectories(standardVersion.listFiles(), false);
+
+                // Note: subdirectories are always pulled from the standard folder path if possible
+                Set<File> fileSet = new TreeSet<>(new FileComparator());
+
+                // Handle FileList
+                fileSet.addAll(fileList);
+                fileSet.addAll(standardFiles);
+                fileList.clear();
+                fileList.addAll(fileSet);
+                fileSet.clear();
+
+                // Handle subdirectoriesList
+                fileSet.addAll(subdirectoryList);
+                fileSet.addAll(userDataSubdirectories);
+                fileList.clear();
+                fileList.addAll(fileSet);
+            } else {
+                // If we do not, then we just return the standard files list based on the files within
+                fileList.addAll(standardFiles);
+            }
+        } else {
+            fileList = new ArrayList<>();
+            if (userDataVersion.exists()) {
+                fileList.add(userDataVersion);
+            } else {
+                fileList.add(standardVersion);
+            }
+        }
+    }
+
+    /**
+     *
+     * @param files an array of files to split between subdirectories and files
+     * @param type true if you want to return files, false for subdirectories
+     * @return a list of either files or subdirectories based on the value of type
+     */
+    private List<File> splitFilesAndSubdirectories(File[] files, boolean type) {
```

Comment:
This is not side effect free and is somewhat confusing when layered with the code calling it above. It is hard to reason about what exactly happens.